pkg/codegen/python/gen.go

```diff
@@ -1302,10 +1302,14 @@ func (mod *modContext) genResource(res *schema.Resource) (string, error) {
 		// type, we must project all properties as strings. For all properties that are not strings,
 		// we'll marshal them to JSON and use the JSON string as a string input.
 		if res.IsProvider && !isStringType(prop.Type) {
-			arg = fmt.Sprintf("pulumi.Output.from_input(%s).apply(pulumi.runtime.to_json) if %s is not None else None", arg, arg)
+			if prop.Secret {
+				arg = fmt.Sprintf("pulumi.Output.from_input(%s).apply(pulumi.runtime.to_json) if %s is not None else None", arg, arg)
+			} else {
+				arg = fmt.Sprintf("pulumi.Output.secret(%s).apply(pulumi.runtime.to_json) if %s is not None else None", arg, arg)
+			}
 		}
 		name := PyName(prop.Name)
-		if prop.Secret {
+		if prop.Secret && isStringType(prop.Type) {
```

Comment:
I think the `&& isStringType(prop.Type)` check is going to break non-Provider resources that have secret properties that aren't strings.

Rather than checking `&& isStringType(prop.Type)`, how about a bool flag whether we already handled the secret? Something like:

```go
var handledSecret bool
if res.IsProvider && !isStringType(prop.Type) {
	if prop.Secret {
		arg = fmt.Sprintf("pulumi.Output.secret(%s).apply(pulumi.runtime.to_json) if %s is not None else None", arg, arg)
		handledSecret = true
	} else {
		arg = fmt.Sprintf("pulumi.Output.from_input(%s).apply(pulumi.runtime.to_json) if %s is not None else None", arg, arg)
 	}
}
name := PyName(prop.Name)
if prop.Secret && !handledSecret {
	fmt.Fprintf(w, "            __props__.__dict__[%[1]q] = None if %[2]s is None else pulumi.Output.secret(%[2]s)\n", name, arg)
} else {
 	fmt.Fprintf(w, "            __props__.__dict__[%q] = %s\n", name, arg)
}
```